### ISSUE TITLE
Trouble understanding subsystem indexes

### ISSUE THREAD

**Issue by ntwalibas** [vsoftco/qpp#53]
Kindly consider the following example:

```C++
qpp::ket ket_1 = 1_ket;
qpp::ket ket_2 = 0_ket;
qpp::ket ket_3 = 0_ket;

// take the tensor product of ket_1 and ket_2
qpp::ket result_1 = qpp::kron(ket_1, ket_2);

// take the tensor product again of result_1 and ket_3
qpp::ket result_2 = qpp::kron(result_1, ket_3);

// take the tensor product of ket_3 and result_1
qpp::ket result_3 = qpp::kron(ket_3, result_1);
```
What I wish to know are the indices of the different subsystems after the application of each tensor product:
- In `result_1`, what is the index of `ket_1` and of `ket_2`:
- In `result_2`, what is the index of `ket_1`, `ket_2` and `ket_3`?
- In `result_3`, what is the index of `ket_1`, `ket_2` and `ket_3`?

What is the general rule to follow to find the indices?
Once those indices are found, what is the right way to use them? For instance, consider the following:
```C++
qpp::ket ket_1 = 1_ket;
qpp::ket ket_2 = 0_ket;
qpp::ket ket_3 = 0_ket;

// take the tensor product of ket_1 and ket_2
qpp::ket result_1 = qpp::kron(ket_1, ket_2);
// is this the correct way to apply the hadamard gate to ket_2 in result_1? 
result_1 = qpp::apply(result_1, gt.H, {ket_2_index});

// take the tensor product again of result_1 and ket_3
qpp::ket result_2 = qpp::kron(result_1, ket_3);
// Is the line below the correct way to apply controlled NOT with ket_1 as control and ket_3 as target?
qpp::ket result_2 = qpp::applyCTRL(result_2, gt.X, {ket_1_index}, {ket_3_index});
// Or this the correct way?
qpp::ket result_2 = qpp::applyCTRL(result_2, gt.X, {0}, {1});
```
This doesn't seem to be documented in the reference manual or the wiki.  

Help again greatly appreciated.

**Comment by vsoftco** [comment]
The indices are in lexicographical order. Say that you have 3 kets, |a>, |b> and |c>. So if you take the Kronecker product of |a> and |b>, then again with |c>, i.e. kron(kron(a,b), c), you get |abc>. The index of |a> is 0, of |b> is 1, and of |c> is 2. Also, if you do kron(c, kron(a,b)), you get |cab>, and the index of |c> is 0, the index of |a> is 1, and the index of |b> is 2. So it really is lexicographical order.

In your example, result_1 = |10>, and result_2 = |100>.  To apply CNOT with ket_1 as control and ket_3 as target, then you need to do `qpp::applyCTRL(result_2, gt.X, {0}, {2});`, as ket_3 will have index 2. Hope that helps.

**Comment by ntwalibas** [comment]
Yes, that helps a lot, thank you! I was able to figure it out but it is always best when the original author confirms hypotheses him/herself.  
While QC books work in lexicographic order, many tools (Qiskit, QUIL,...) are adopting the computer science convention of counting backwards.

That said, Quantum++ has been of immense help. I've got myself a prototype quantum programming language working thanks to QPP. Please do check it out at [https://github.com/avalon-lang/avaloniq](https://github.com/avalon-lang/avaloniq). Hopefully, one day it will feature among projects using this library.

Cheers!

**Comment by ntwalibas** [comment]
I find myself in unchartered territory again with subsystem indices.  
I have a situation similar to the one below:

```C++
ket a = 1_ket;
ket b = 0_ket;

ket res = kron(kron(a, b), kron(a, b));
res = applyCTRL(res, gt.CNOT, {a_index}, {b_index});
```

The tensor product gives an object of the type |abab>. Reaching code like the one above threw a wrench in the scheme: what index do I assign to `a` and which one to `b`? I could use 0 or 2 for `a` and 1 or 3 for `b`. But not both.

What would be the idiomatic way to solve this problem with Quantum++?
Thanks again for your time.

**Comment by vsoftco** [comment]
Not sure what you try to achieve? You want something like a Controlled-Controlled-X-X? In that case, just do `applyCTRL(res, gt.X, {0,2}, {1,3})`.

**Comment by ntwalibas** [comment]
Yeah, something like that!  
My case was something like `cnot(a, b); cnot(b, a)` which ended up generating the tensor product above. The first CNOT generates the tensor product `kron(a, b)` (that is before calling `applyCTRL`). The resulting state is recorded for both `a` and `b` with their indices recorded. The second CNOT takes the preceding state from `kron(a, b)` and tries to create `kron(kron(a, b), kron(a, b))`. I could just apply the second CNOT on the existing tensor product but I needed to detect when I could forgo doing a tensor product and the method I found was not reliable.

I ended up using early measurements. Though I don't like the solution, it *seems* to work for now.  
I expect it to eventually fail for some corner case I haven't encountered yet and I'm happy with the solution you just proposed. I will test it before things get bad.

Again, thank you very much.
